docs/jotai/advanced-recipes/atom-creators.mdx

```diff
@@ -0,0 +1,82 @@
+---
+title: Atom creators
+nav: 5.2
+---
+
+## `atomWithToggle`
+
+> `atomWithToggle` creates a new atom with a boolean as initial state & a setter function to toggle it.
+
+This avoids the boilerplate of having to setup another atom just to update the state of the first.
+
+```ts
+export function atomWithToggle(
+  initialValue?: boolean,
+  createAtom: (initialValue: boolean) => PrimitiveAtom<boolean> = atom
```

Comment:
This is from my suggestion, but there was a misunderstanding of mine.
Looking at the implementation, the default `atom` is useless.

May I suggest something completely different?
- A recipe for `atomWithToggle` without createAtom (it's very simple one)
- A recipe for `atomWithToggleAndStorage` (basically your original idea. `key` doesn't have to be optional for a recipe, or it's another recipe.)

Reply:
FWIW, I was expecting something like this, but it's an extra abstraction if we don't use `atomWithStorage` for `createAtom`.

```ts
export function atomWithToggle(
  initialValue?: boolean,
  createAtom: (initialValue: boolean) => PrimitiveAtom<boolean> = atom
): WritableAtom<boolean, boolean | undefined> {
  const anAtom = createAtom(initialValue)
  const derivedAtom = atom(
    (get) => get(anAtom),
    (get, set, nextValue?: boolean) => {
      const update = state ?? !get(anAtom)
      set(anAtom, update)
    }
  )
  return derivedAtom
}
```

For non-optional atomWithToogleAndStorage, this derivedAtom style might be better.

Reply:
I made changes using this style, I hope that's what you expected this time, please tell me otherwise !